File: typescript/src/extras/mcp.ts
/**
 * MCP (Model Context Protocol) integration for Asqav.
 *
 * Install (peer dependency):
 *   npm install @modelcontextprotocol/sdk
 *
 * Usage:
 *   import { init } from "@asqav/sdk";
 *   import { enableMcpGovernance } from "@asqav/sdk/extras/mcp";
 *   import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
 *
 *   init({ apiKey: process.env.ASQAV_API_KEY! });
 *   const server = new McpServer({ name: "my-server", version: "1.0.0" });
 *   enableMcpGovernance(server, { agentName: "my-mcp-server" });
 *   server.registerTool("add", { description: "add" }, async ({ a, b }) => ({
 *     content: [{ type: "text", text: String(a + b) }],
 *   }));
 *
 * One call to ``enableMcpGovernance`` turns governance on for the whole
 * server: every tool call signs an Asqav receipt by default, regardless of
 * whether the tool was registered before or after the call, and a tool that
 * throws signs ``tool:error``. Signing is fail-open. This module never
 * imports ``@modelcontextprotocol/sdk`` (it duck-types the passed server),
 * so the peer stays optional and importing without calling changes nothing.
 */

import { AsqavAdapter, defaultOnError, type AsqavAdapterOptions } from "./_base.js";

// Boundary type for the duck-typed server. ``any`` rest params let a
// concretely-typed McpServer.registerTool/tool satisfy this shape.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ToolRegistrar = (...args: any[]) => any;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyFn = (...args: any[]) => any;

/** Minimal server shape: a ``_registeredTools`` registry, or registrar-only (registerTool/tool). */
export interface McpServerLike {
  _registeredTools?: Record<string, unknown>;
  registerTool?: ToolRegistrar;
  tool?: ToolRegistrar;
}

export type AsqavMcpAdapterOptions = AsqavAdapterOptions;

const WRAP_FLAG = "__asqavWrapped";
// Symbol key so the funnel marker never shows up in Object.keys/entries
// (McpServer enumerates _registeredTools to build tools/list).
const FUNNEL_FLAG = Symbol.for("asqav.mcp.funnel");
// Dispatch-time handler field: ``handler`` on recent SDK versions,
// ``callback`` on older ones. Wrap whichever is present.
const HANDLER_KEYS = ["handler", "callback"] as const;
const MAX_ERROR_LEN = 200;

/** Signs ``tool:call`` per MCP tool call and ``tool:error`` on a throw, fail-open and order-independent. */
export class AsqavMcpAdapter extends AsqavAdapter {
  /** Wire signing into ``server``'s tool dispatch in place; returns ``server``. */
  instrument<T extends McpServerLike>(server: T): T {
    const registry = server._registeredTools;
    if (registry && typeof registry === "object") {
      this.wrapDispatchRegistry(server);
      return server;
    }
    const hasRegister = typeof server.registerTool === "function";
    const hasTool = typeof server.tool === "function";
    if (!hasRegister && !hasTool) {
      throw new Error(
        "enableMcpGovernance expects an MCP server exposing a tool registry, "
        + "registerTool or tool (e.g. McpServer from @modelcontextprotocol/sdk).",
      );
    }
    if (hasRegister) this.wrapRegistrar(server, "registerTool");
    if (hasTool) this.wrapRegistrar(server, "tool");
    return server;
  }

  // Single dispatch funnel: every tools/call lookup reads the registry, so a
  // Proxy get that wraps the handler signs order-independently.
  private wrapDispatchRegistry(server: McpServerLike): void {
    const target = server._registeredTools as Record<string | symbol, unknown>;
    if (target[FUNNEL_FLAG]) return;
    const adapter = this;

    const proxy = new Proxy(target, {
      get(t, prop, receiver): unknown {
        const value = Reflect.get(t, prop, receiver);
        if (typeof prop !== "string" || value === null || typeof value !== "object") {
          return value;
        }
        const tool = value as Record<string, unknown>;
        let wrapped: Record<string, unknown> | null = null;
        for (const key of HANDLER_KEYS) {
          if (typeof tool[key] === "function") {
            // Shallow copy keeps enabled/schemas/update intact; the stored
            // tool stays unwrapped so re-reads never double-wrap.
            wrapped = wrapped ?? { ...tool };
            wrapped[key] = adapter.wrapToolCallback(prop, tool[key] as AnyFn);
          }
        }
        return wrapped ?? value;
      },
    });

    target[FUNNEL_FLAG] = true;
    server._registeredTools = proxy as Record<string, unknown>;
  }

  private wrapRegistrar(server: McpServerLike, method: "registerTool" | "tool"): void {
    const original = server[method];
    if (!original) return;
    if ((original as unknown as Record<string, unknown>)[WRAP_FLAG]) return;
    const adapter = this;

    const wrapped = function (this: unknown, ...args: unknown[]): unknown {
      const name = typeof args[0] === "string" ? args[0] : "unknown";
      // The tool callback is the last function argument in every overload.
      let cbIndex = -1;
      for (let i = args.length - 1; i >= 0; i -= 1) {
        if (typeof args[i] === "function") {
          cbIndex = i;
          break;
        }
      }
      if (cbIndex >= 0) {
        args[cbIndex] = adapter.wrapToolCallback(name, args[cbIndex] as AnyFn);
      }
      return (original as ToolRegistrar).apply(this, args);
    };
    (wrapped as unknown as Record<string, unknown>)[WRAP_FLAG] = true;
    server[method] = wrapped as ToolRegistrar;
  }

  // Shared per-call wrap: sign tool:call, run the tool, sign tool:error on a
  // sync throw or an async rejection, then rethrow.
  private wrapToolCallback(name: string, originalCb: AnyFn): AnyFn {
    if ((originalCb as unknown as Record<string, unknown>)[WRAP_FLAG]) return originalCb;
    const adapter = this;

    const wrapped = function (this: unknown, ...cbArgs: unknown[]): unknown {
      adapter.emitToolCall(name, cbArgs[0]);
      let out: unknown;
      try {
        out = originalCb.apply(this, cbArgs);
      } catch (err) {
        adapter.emitToolError(name, err);
        throw err;
      }
      if (out instanceof Promise) {
        return out.catch((err: unknown) => {
          adapter.emitToolError(name, err);
          throw err;
        });
      }
      return out;
    };
    (wrapped as unknown as Record<string, unknown>)[WRAP_FLAG] = true;
    return wrapped;
  }

  /** Sign a governance receipt for one MCP tool call (fail-open). */
  emitToolCall(name: string, args: unknown): void {
    const argKeys = args && typeof args === "object"
      ? Object.keys(args as Record<string, unknown>)
      : [];
    this.signAction({
      actionType: "tool:call",
      context: { tool: name, arg_keys: argKeys },
    });
  }

  /** Sign tool:error when a wrapped tool throws (fail-open). */
  emitToolError(name: string, err: unknown): void {
    // String(err) can itself throw (e.g. Object.create(null)). Keep the
    // stringification inside the fail-open boundary so it can never mask
    // the tool's own error.
    try {
      const errorType = err instanceof Error
        ? err.constructor.name
        : typeof err;
      const message = err instanceof Error ? err.message : String(err);
      this.signAction({
        actionType: "tool:error",
        context: {
          tool: name,
          error_type: errorType,
          error: message.slice(0, MAX_ERROR_LEN),
        },
      });
    } catch (emitErr) {
      const onError = this.options.onError ?? defaultOnError;
      try {
        onError(emitErr, { actionType: "tool:error" });
      } catch {
        // Swallow secondary errors from the user-supplied handler.
      }
    }
  }
}

/** Default-on Asqav governance: every tool call signs a receipt, even tools registered before this call. */
export function enableMcpGovernance<T extends McpServerLike>(
  server: T,
  opts: AsqavMcpAdapterOptions,
): T {
  const adapter = new AsqavMcpAdapter(opts);
  return adapter.instrument(server);
}

File: python/src/asqav/extras/mcp.py
"""MCP (Model Context Protocol) integration for asqav.

Install: pip install asqav[mcp]

Usage::

    from mcp.server.fastmcp import FastMCP
    from asqav.extras.mcp import enable_mcp_governance

    server = FastMCP("my-server")
    enable_mcp_governance(server, agent_name="my-mcp-server")

    @server.tool()
    def add(a: int, b: int) -> int:
        return a + b

One call to ``enable_mcp_governance`` turns governance on for the whole
server: every tool call signs an asqav receipt by default, with no
per-call flag. Signing is fail-open - a signing error is logged and never
blocks or alters the tool call. Importing this module without calling the
function changes nothing, and ``mcp`` is an optional extra (this module
duck-types the server, it never imports ``mcp`` itself).
"""

from __future__ import annotations

import inspect
import logging
from typing import Any

from ._base import AsqavAdapter

logger = logging.getLogger("asqav")

__all__ = ["AsqavMCPAdapter", "enable_mcp_governance"]

_MAX_LEN = 200


class AsqavMCPAdapter(AsqavAdapter):
    """Signs an asqav governance receipt on every MCP tool call.

    ``instrument`` wraps the server's tool-dispatch funnel so each call
    records ``tool:call`` (and ``tool:error`` when the tool raises). It
    handles a FastMCP server (wraps the ``_tool_manager`` dispatch that both
    the direct API and real request routing funnel through) and a low-level
    ``Server`` (wraps the ``call_tool`` decorator so registered handlers
    sign). Fail-open throughout.

    Args:
        api_key: Optional API key override (uses ``asqav.init()`` default).
        agent_name: Name for an asqav agent (calls ``Agent.create``).
        agent_id: ID of an existing asqav agent (calls ``Agent.get``).
        observe: When True, log intent instead of signing.
    """

    def instrument(self, server: Any) -> Any:
        """Wire signing into ``server``'s tool dispatch; returns ``server``."""
        tool_manager = getattr(server, "_tool_manager", None)
        if tool_manager is not None and inspect.iscoroutinefunction(
            getattr(tool_manager, "call_tool", None)
        ):
            self._wrap_dispatch(tool_manager, "call_tool")
            return server
        if inspect.iscoroutinefunction(getattr(server, "call_tool", None)):
            self._wrap_dispatch(server, "call_tool")
            return server
        if callable(getattr(server, "call_tool", None)):
            self._wrap_lowlevel_decorator(server)
            return server
        raise TypeError(
            "enable_mcp_governance expects a FastMCP server or a low-level mcp "
            f"Server exposing call_tool; got {type(server).__name__}."
        )

    def _wrap_dispatch(self, target: Any, attr: str) -> None:
        # Single async funnel: wrap once so both direct calls and routing sign.
        original = getattr(target, attr)
        if getattr(original, "_asqav_wrapped", False):
            return
        adapter = self

        async def _signed_dispatch(
            name: str, arguments: Any, *args: Any, **kwargs: Any
        ) -> Any:
            await adapter._sign_tool_call(name, arguments)
            try:
                return await original(name, arguments, *args, **kwargs)
            except Exception as exc:
                await adapter._sign_tool_error(name, exc)
                raise

        _signed_dispatch._asqav_wrapped = True  # type: ignore[attr-defined]
        setattr(target, attr, _signed_dispatch)

    def _wrap_lowlevel_decorator(self, server: Any) -> None:
        # Low-level Server.call_tool is a decorator; wrap the handler it registers.
        original_call_tool = server.call_tool
        if getattr(original_call_tool, "_asqav_wrapped", False):
            return
        adapter = self

        def _patched_call_tool(*d_args: Any, **d_kwargs: Any) -> Any:
            real_decorator = original_call_tool(*d_args, **d_kwargs)

            def _wrapping_decorator(func: Any) -> Any:
                async def _signed_handler(
                    name: str, arguments: Any, *a: Any, **k: Any
                ) -> Any:
                    await adapter._sign_tool_call(name, arguments)
                    try:
                        return await func(name, arguments, *a, **k)
                    except Exception as exc:
                        await adapter._sign_tool_error(name, exc)
                        raise

                return real_decorator(_signed_handler)

            return _wrapping_decorator

        _patched_call_tool._asqav_wrapped = True  # type: ignore[attr-defined]
        server.call_tool = _patched_call_tool

    async def _sign_tool_call(self, name: str, arguments: Any) -> None:
        """Sign tool:call for one MCP tool invocation (fail-open)."""
        try:
            arg_keys = sorted(arguments.keys()) if isinstance(arguments, dict) else []
            await self._sign_action_async(
                "tool:call",
                {"tool": str(name), "arg_keys": arg_keys},
            )
        except Exception as exc:
            logger.warning("asqav mcp tool:call signing failed (fail-open): %s", exc)

    async def _sign_tool_error(self, name: str, exc: BaseException) -> None:
        """Sign tool:error when a wrapped tool raises (fail-open)."""
        try:
            await self._sign_action_async(
                "tool:error",
                {
                    "tool": str(name),
                    "error_type": type(exc).__name__,
                    "error": str(exc)[:_MAX_LEN],
                },
            )
        except Exception as sign_exc:
            logger.warning(
                "asqav mcp tool:error signing failed (fail-open): %s", sign_exc
            )


def enable_mcp_governance(
    server: Any,
    *,
    api_key: str | None = None,
    agent_name: str | None = None,
    agent_id: str | None = None,
    observe: bool = False,
) -> AsqavMCPAdapter:
    """Turn on default-on asqav governance for an MCP server.

    After this one call, every tool call on ``server`` signs a governance
    receipt. The server is instrumented in place; the returned adapter holds
    the recorded signatures.
    """
    adapter = AsqavMCPAdapter(
        api_key=api_key,
        agent_name=agent_name or "asqav-mcp-server",
        agent_id=agent_id,
        observe=observe,
    )
    adapter.instrument(server)
    return adapter
